django/db/models/fields/files.py

```diff
@@ -374,12 +383,20 @@ def __set__(self, instance, value):
         # Model.__init__().  The ImageField.update_dimension_fields method
         # hooked up to the post_init signal handles the Model.__init__() cases.
         # Assignment happening outside of Model.__init__() will trigger the
-        # update right here.
+        # update using ImageFile._set_dimensions
         if previous_file is not None:
-            self.field.update_dimension_fields(instance, force=True)
+
+            # If the original value passed in was a string, it's usually necessary
+            # to convert it to a file object, here done using the descriptor's
+            # __get__ method.
+            if isinstance(value, str):
+                value = getattr(instance, self.field.attname)
+
+            self.field._set_dimensions(instance, value)
```

Comment:
I can revert this and the tests still pass

Reply:
I'm not exactly sure what you want me to do with this. This is a very tricky code path. What would a test look like?

Reply:
I'm trying to establish the smallest changeset to achieve the desired result, which we will likely squash into a single commit 🤔 
As you have a test checking the extra read isn't happening, do we need these changes?

Reply:
Yes, the extra read is covered by `NoReadTests`. If you port the `NoReadTests` back to main/trunk, you'll see that they fail.

Reply:
If you go look at the signature of `_set_dimensions` you'll see that the required type of `value` is a file-like object or `None`. That's what is has to be coerced here.

Reply:
Ah I'm not sure if I was clear, I think we can revert this specific bit to what it was before 
```python
self.field.update_dimension_fields(instance, force=True)
```
I see you have updated `update_dimension_fields` to use `_set_dimensions` and when I do this change, the test still passes